doc/source/how-to/manage-protobuf-definitions.rst

```diff
@@ -0,0 +1,163 @@
+Manage Protobuf definitions
+===========================
+
+Protobuf service definitions provide the API specification for underlying
+server implementations, so that each consuming client library has a clear
+contract for gRPC data messages. Ideally, the .proto files have a single
+repository established as the source of truth for the .proto files,
+organized by API version increment as the API definition expands and changes.
+Since most client libraries are custom implementations enhancing the developer
+experience when consuming the service, releasing the Protobuf definitions
+publicly gives full flexibility to the developer to operate at the abstraction
+layer they choose to do so.
+
+Maintain API definition repository
+----------------------------------
+
+The Protobuf definition of the service is language agnostic, so the repository
+containing the Protobuf files can be created within the top-level Ansys
+GitHub organization. Every update of the Protobuf files follows a standard
+pull request process as a sanity-check for API definition accuracy. Language
+specific packages can be generated for each merge or on a set cadence.
+
+Managing Protobuf definitions for Python clients
+------------------------------------------------
+
+Building Python stub classes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Add a python project definition to the API repository that can be utilized
+within the CI/CD build pipeline.
+
+A common tool to help build python stub classes is available and maintained
+within Ansys:
+
+`ansys-tools-protoc-helper <https://github.com/ansys/ansys-tools-protoc-helper/>`_
+
+Example inclusion within pyproject.toml build dependency configuration:
+
+.. code-block:: toml
+
+    [build-system]
+    requires = ["setuptools >= 42.0.0", "wheel", "ansys_tools_protoc_helper"]
+
```

Comment:
The projects I've seen all have a setup.py file that either supersedes or works in conjunction with pyproject.toml. The template https://github.com/ansys/ansys-api-template creates setup.py. I inadvertently found out that Python 3.11 no longer supports setup.py (or strongly discourages its use). At some point pyproject.toml will need to be much more robust. Rather than including the example here, you might just refer folks to https://github.com/ansys/ansys-api-template, which, I assume, will be updated when setup.py is no longer used.

Reply:
In the `ansys-api-template`, the use of `setup.py` [1] is unavoidable. The reason is that the way we generate the `.py` files from `.proto` ties into the steps of the `setuptools` - based build. In a sense, this is similar to having compiled code in your package, where you also need control over how that compilation is done. More details here: https://github.com/ansys/ansys-tools-protoc-helper#manual-use

Can you elaborate on how Python 3.11 no longer supports `setup.py`? As far as I'm aware, there's still a rather large number of projects doing setuptools-based builds. Most notably, numpy: https://github.com/numpy/numpy/blob/main/setup.py

[1] or alternatively, `setup.cfg`; in any case, it needs to be a setuptools-based build.

Reply:
@greschd I saw errors like the following.
```
2022-10-26T17:34:13.4966702Z Installing collected packages: six, protobuf, numpy, grpcio, grpcio-tools
2022-10-26T17:34:16.1597729Z   DEPRECATION: grpcio is being installed using the legacy 'setup.py install' method, because it does not have a 'pyproject.toml' and the 'wheel' package is not installed. pip 23.1 will enforce this behaviour change. A possible replacement is to enable the '--use-pep517' option. Discussion can be found at https://github.com/pypa/pip/issues/8559
2022-10-26T17:34:16.1599084Z   Running setup.py install for grpcio: started
2022-10-26T17:35:16.2974285Z   Running setup.py install for grpcio: still running...
2022-10-26T17:36:16.3927973Z   Running setup.py install for grpcio: still running...
2022-10-26T17:37:10.8230847Z   Running setup.py install for grpcio: finished with status 'error'
2022-10-26T17:37:11.2571113Z   error: subprocess-exited-with-error
```

This could very well be a misunderstanding on my part. It could just be the result of `grpcio` and other libraries, like `numpy`, not having a wheel available for 3.11.

Reply:
That seems mostly because 3.11 is new, and those projects might not have had a release including pre-built wheels for 3.11 yet. [1] 

Typically one can still build from the source distribution (sdist), but that may fail due e.g. to missing compilers, or other missing built-time dependencies.

Once a wheel is created, as far as I know the "build backend" (setuptools, flit, poetry) is no longer involved; the package has been baked into a standardized format.

[1] For `numpy`, the first release with 3.11 wheels appears to be `1.23.2` (look for `*cp311_*.whl` in [this list](https://pypi.org/project/numpy/1.23.2/#files)). For `grpcio`, it's [`1.49.1`](https://pypi.org/project/grpcio/1.49.1/#files).